### PR TITLE
[DNM] Field of View component [bounty]

### PR DESCRIPTION
**Pull request by Azarak**
## About The Pull Request

DNM AS I'M DOING IT UPSTREAM

Some notes of mine:
-In combat mode you will "stick" to a direction when you face something by clicking on it for 2 seconds, preventing movements to change it. You can disable it in prefs -> gameplay. This is so you can keep looking into one direction in combat without any extra keybinds in play.

Some ideas and the mask assets were taken from https://github.com/Citadel-Station-13/Citadel-Station-13/pull/12075
_____________

Hey, host here. I asked my friend Azarak here to work on this as a bounty. We've changed the design, so that it offers a much less intrusive and disorienting experience. First of all, the server can toggle field of vision in-game or in config. 

Next, everything is visible save a player in combat mode. Objects will not be invisible, as that causes disorientation. All players that are not in combat mode are always visible. This makes it so that it does not impact normal roleplay. If a player enters combat mode, they then become invisible to a person if they're not in said person's field of vision. Furthermore, if a player is not in combat mode, the vision cone is not visible which should cause minimal problems, though this can be changed in preferences to see a reduced alpha vision cone when not in combat mode if a player wants visual cues on the limits of their field of vision. For many, this should not disrupt their gameplay and cause minimal disorientation if any.

Vision cones are set to at 180 degrees, though this may be increased based on tests. Also, the player can see one tile back as well to see the player they're dragging behind them. Lastly, we will be adding visual effects which is not yet in. For instance, if a player is not walking or crawling, white footstep markers will give them away. There will also be visual effects for attacking, i.e. shooting or melee as well.

This should add an interesting element to stealth without being burdensome to gameplay.

https://user-images.githubusercontent.com/6381979/144940526-25b2952f-e919-4b5e-8627-96d2c4be6f61.mp4

## How This Contributes To The Skyrat Roleplay Experience
epic metal gear solid gameplay

## Changelog

:cl:
add: Added a Field of View.
/:cl:

### CODE REVIEW

File: code/modules/projectiles/projectile.dm [hunk 1]
```diff
@@ -12,6 +12,7 @@
 	wound_bonus = CANT_WOUND // can't wound by default
 	generic_canpass = FALSE
 	blocks_emissive = EMISSIVE_BLOCK_GENERIC
+	plane = GAME_PLANE_FOV_HIDDEN
```
Comment (Gandalf2k15) [inline code/modules/projectiles/projectile.dm]:
```suggestion
	plane = GAME_PLANE_FOV_HIDDEN //SKYRAT EDIT ADDITION
```

File: code/modules/client/preferences/ambient_occlusion.dm [hunk 2]
```diff
@@ -10,3 +10,10 @@
 		return
 
 	plane_master.backdrop(client.mob)
+
+	var/atom/movable/screen/plane_master/game_world/fov_hidden/plane_master_fov = locate() in client?.screen
+	if (!plane_master_fov)
+		return
+
+	plane_master_fov.backdrop(client.mob)
```
Comment (Gandalf2k15) [inline code/modules/client/preferences/ambient_occlusion.dm:19]:
```suggestion
	//SKYRAT EDIT ADDITION
	var/atom/movable/screen/plane_master/game_world/fov_hidden/plane_master_fov = locate() in client?.screen
	if (!plane_master_fov)
		return

	plane_master_fov.backdrop(client.mob)
	//SKYRAT EDIT END
```

File: code/game/objects/effects/temporary_visuals/temporary_visual.dm [hunk 3]
```diff
@@ -3,6 +3,7 @@
 	icon_state = "nothing"
 	anchored = TRUE
 	layer = ABOVE_MOB_LAYER
+	plane = GAME_PLANE_FOV_HIDDEN
```
Comment (Gandalf2k15) [inline code/game/objects/effects/temporary_visuals/temporary_visual.dm]:
```suggestion
	plane = GAME_PLANE_FOV_HIDDEN // SKYRAT EDIT ADDITION
```

File: code/game/objects/effects/temporary_visuals/projectiles/projectile_effects.dm [hunk 4]
```diff
@@ -3,6 +3,7 @@
 	icon = 'icons/obj/guns/projectiles.dmi'
 	icon_state = "nothing"
 	layer = ABOVE_MOB_LAYER
+	plane = GAME_PLANE_FOV_HIDDEN
```
Comment (Gandalf2k15) [inline code/game/objects/effects/temporary_visuals/projectiles/projectile_effects.dm]:
```suggestion
	plane = GAME_PLANE_FOV_HIDDEN // SKYRAT EDIT ADDITION
```

File: code/game/atoms_movable.dm [hunk 5]
```diff
@@ -80,6 +80,8 @@
 	var/contents_thermal_insulation = 0
 	/// The degree of pressure protection that mobs in list/contents have from the external environment, between 0 and 1
 	var/contents_pressure_protection = 0
+	/// Time until we stop "sticky" facing towards something. Set on face_atom() if someone with combat mode faces
+	var/sticky_facing_until = 0
```
Comment (Gandalf2k15) [inline code/game/atoms_movable.dm:85]:
```suggestion
	// SKYRAT EDIT ADDITION
	/// Time until we stop "sticky" facing towards something. Set on face_atom() if someone with combat mode faces
	var/sticky_facing_until = 0
	// SKYRAT EDIT END
```

File: code/game/atoms_movable.dm [hunk 6]
```diff
@@ -387,7 +389,7 @@
 	if(!direction)
 		direction = get_dir(src, newloc)
 
-	if(set_dir_on_move)
+	if(set_dir_on_move && sticky_facing_until < world.time)
```
Comment (Gandalf2k15) [inline code/game/atoms_movable.dm:453]:
```suggestion
	if(set_dir_on_move && sticky_facing_until < world.time) //SKYRAT EDIT CHANGE
```

File: code/game/atoms_movable.dm [hunk 7]
```diff
@@ -514,7 +516,7 @@
 						moving_diagonally = SECOND_DIAG_STEP
 						. = step(src, SOUTH)
 			if(moving_diagonally == SECOND_DIAG_STEP)
-				if(!. && set_dir_on_move)
+				if(!. && set_dir_on_move && sticky_facing_until < world.time)
```
Comment (Gandalf2k15) [inline code/game/atoms_movable.dm]:
```suggestion
				if(!. && set_dir_on_move && sticky_facing_until < world.time) //SKYRAT EDIT CHANGE
```

File: code/game/atoms_movable.dm [hunk 8]
```diff
@@ -546,7 +548,7 @@
 
 	last_move = direct
 
-	if(set_dir_on_move)
+	if(set_dir_on_move && sticky_facing_until < world.time)
```
Comment (Gandalf2k15) [inline code/game/atoms_movable.dm]:
```suggestion
	if(set_dir_on_move && sticky_facing_until < world.time) //SKYRAT EDIT CHANGE
```

File: code/_onclick/hud/rendering/plane_master.dm [hunk 9]
```diff
@@ -72,6 +72,13 @@
 	if(istype(mymob) && mymob.client?.prefs?.read_preference(/datum/preference/toggle/ambient_occlusion))
 		add_filter("AO", 1, drop_shadow_filter(x = 0, y = -2, size = 4, color = "#04080FAA"))
 
+/atom/movable/screen/plane_master/game_world/fov_hidden
```
Comment (Gandalf2k15) [inline code/_onclick/hud/rendering/plane_master.dm:75]: Stuff added in this file can probably go in modular_skyrat/master_files/code/_onclick/hud/rendering/plane_master.dm

File: code/__DEFINES/layers.dm [hunk 10]
```diff
@@ -112,6 +116,9 @@
 ///Things that should render ignoring lighting
 #define ABOVE_LIGHTING_PLANE 120
 
+///Plane for sounds creating visual indicators
```
Comment (Gandalf2k15) [inline code/__DEFINES/layers.dm]: Add dis stuff into it's own file in the skyrat defines

File: code/__DEFINES/dcs/signals/signals_mob/signals_mob_living.dm [hunk 11]
```diff
@@ -77,3 +77,7 @@
 #define COMSIG_LIVING_SLAM_TABLE "living_slam_table"
 ///(NOT on humans) from mob/living/*/UnarmedAttack(): (atom/target, proximity, modifiers)
 #define COMSIG_LIVING_UNARMED_ATTACK "living_unarmed_attack"
+
+//SKYRAT LIVING SIGNALS
+/// From mob/living/*/set_combat_mode(): (new_state)
+#define COMSIG_LIVING_COMBAT_MODE_TOGGLE "living_combat_mode_toggle"
```
Comment (Gandalf2k15) [inline code/__DEFINES/dcs/signals/signals_mob/signals_mob_living.dm:83]: Can go in it's own file in skyrat defines

**Review by Gandalf2k15 — CHANGES_REQUESTED** [review]
just modular stuff assuming you don't plan on changing it

**Comment by CandleJaxx** [comment]
it'd be cool to see other peoples' vision cones as well, give it a metal gear solid style feeling if you sneak around

**Comment by ZephyrTFA** [comment]
How does this work with binocs and other stuff that moves your clients eye. Such as xenobio consoles

**Comment by Azarak** [comment]
> How does this work with binocs and other stuff that moves your clients eye. Such as xenobio consoles

good point, it doesn't lol. I'll fix that soon

File: modular_skyrat/modules/field_of_view/code/field_of_view.dm [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+/mob/living/Login()
+	. = ..()
+	if(has_field_of_view && CONFIG_GET(flag/fov_enabled))
+		AddComponent(/datum/component/field_of_vision)
+
+/mob/living/Logout()
+	var/fov_component = GetComponent(/datum/component/field_of_vision)
+	if(fov_component)
+		qdel(fov_component)
+	. = ..()
```
Comment (MrMelbert) [inline modular_skyrat/modules/field_of_view/code/field_of_view.dm]: why not hook this onto `COMSIG_MOB_LOGOUT` on the component itself

**Comment by optimumtact** [comment]
interesting

**Comment by sqnztb** [comment]
After playing with it, my only thought is suggesting changing it from combat mode toggling stealth to the CI toggling stealth. It's quite annoying standing around and watching people just disappear when it's clearly not a combat/stealth scenario.

File: modular_skyrat/modules/field_of_view/code/field_of_view.dm (hunk 12 above)
Comment (Azarak) [reply modular_skyrat/modules/field_of_view/code/field_of_view.dm]: it made sense for me since the application of it was in Login, I suppose I could move it

**Comment by Gandalf2k15** [comment]
Changing from draft to PR

**Comment by Azarak** [comment]
but its WIP

**Comment by gamermomentum** [comment]
Without a good mechanic for dictating which direction you're facing, this probably shouldn't see a full-merge. It just makes combat incredibly unwieldy.  
Unless that most recent change fixes it, using binocs permanently owns your FOV by moving your camera so you basically can't see anyone who's sneaking unless they're near the outside of your screen in whatever direction you've binocs'd in. 

Also, this being locked to combat mode and not like a standalone sneak mode is a little dubious, as lots of folks(myself included) use combat mode a LOT to avoid getting help intented/pushed around.

Anyhow, I'd strongly suggest adding a system to have combat mode make your face direction track your cursor, like some other places do. The current "Direction lock for a few seconds on-click" system just doesn't cut it.

**Comment by ChungusGamer666** [comment]
This PR is 🥴 but i'm pretty sure if i go into detail explaining why i'll get muted
All i have to say is 👎 😭😢

**Comment by Higgin** [comment]
> Also, this being locked to combat mode and not like a standalone sneak mode is a little dubious, as lots of folks(myself included) use combat mode a LOT to avoid getting help intented/pushed around.

Would suggest expanding the FOV out to something closer to 225-270 and making the cone always-on rather than tying to combat mode/CI too, because it's very possible to engage in combat without engaging either rn. Makes the significant impact of this highly-inconsistent.

Maybe add a slower 'look around' verb that lets people briefly disable the cone after a visible do_after? (For screenshots and QoL, while also telegraphing that you're looking to folks sneaking up on you.)

Also, this might have broken locker/crate hiding overlays at present - they do not show up while inside on or off CM.

File: code/__DEFINES/layers.dm (hunk 10 above)
Comment (Azarak) [reply code/__DEFINES/layers.dm]: I dont think it's wise to seperate layer and plane definitions

**Comment by Stalkeros2** [comment]
Combat mode is already "oversaturated" with functions and is crucial in X or Y department's/person's gameplay (I myself just randomly flip it on or off when I am very bored so you know that might result in "Schrödinger's people" existing. Besides, mining usually involves changing your intents constantly. And I'm a miner. Flickering miner.), so setting it to the combat mode isn't the best idea tbh.
I'd recommend making a separate stealth mode, just like me and some other folks in Discord had suggested.

**Comment by softcerv** [comment]
> Combat mode is already "oversaturated" with functions and is crucial in X or Y department's/person's gameplay (I myself just randomly flip it on or off when I am very bored so you know that might result in "Schrödinger's people" existing. Besides, mining usually involves changing your intents constantly. And I'm a miner. Flickering miner.), so setting it to the combat mode isn't the best idea tbh.
I'd recommend making a separate stealth mode, just like me and some other folks in Discord had suggested.

having this as like a stealth mode thing that would slow your speed down would be a really good way of handling it ngl. Ambushes would still be possible but you'd also put yourself at far more risk doing so

**Comment by chefferz** [comment]
Love how absolutely nobody is commenting how they really feel because they're worried they'll get banned or muted. And how huge game changes like this are "tested" on a whim without it seemingly being talked about or even suggested, just "hey this is cool" without any actual take on how it will majorly affect game play.

Absolutely 100% not a fan, and have no intent on playing Skyrat while this is implemented. This isn't even remotely needed in a 2D game, and only further adds to the pool of what you'd have to micromanage.

**Comment by chefferz** [comment]
Love how absolutely nobody is commenting how they really feel because they're worried they'll get banned or muted. And how huge game changes like this are "tested" on a whim without it seemingly being talked about or even suggested, just "hey this is cool" without any actual take on how it will majorly affect game play. 

Have there been any actual polls of the BROAD PLAYER BASE versus the loud, minor minority that are in the discord?

Absolutely 100% not a fan, and have no intent on playing Skyrat while this is implemented. This isn't even remotely needed in a 2D game, and only further adds to the pool of what you'd have to micromanage.

Actual raw critique is perfect as mentioned above, have this separate. We really need a server that isn't blatantly messed with "features" that nobody asked for, though, and that doesnt focus on antag satisfaction more than the entire populace.

**Comment by chizzyyyy** [comment]
Apart from the obvious "disappearing people" topic people brought up, I feel as though this is unnecessary bloat to a feature that's already a bit finnicky to get to play with correctly. Having to constantly face your target as they strafe around you will only feel detrimental to the combat experience, I can not see a single scenario this would not be an actively detrimental feature. It will make fights in open spaces very bad, and fights in maints even worse if there's a T junction.
Please do not merge this, I don't think It's a good idea and people most likely agree.

**Comment by chefferz** [comment]
> > Love how absolutely nobody is commenting how they really feel because they're worried they'll get banned or muted. And how huge game changes like this are "tested" on a whim without it seemingly being talked about or even suggested, just "hey this is cool" without any actual take on how it will majorly affect game play.
> > Have there been any actual polls of the BROAD PLAYER BASE versus the loud, minor minority that are in the discord?
> > Absolutely 100% not a fan, and have no intent on playing Skyrat while this is implemented. This isn't even remotely needed in a 2D game, and only further adds to the pool of what you'd have to micromanage.
> > Actual raw critique is perfect as mentioned above, have this separate. We really need a server that isn't blatantly messed with "features" that nobody asked for, though, and that doesnt focus on antag satisfaction more than the entire populace.
> 
> I disagree with 90% of everything you say, including 90% of this comment (what the hell do you mean antag satisfaction, this affects everyone), but this is not a good implementation of field of vision and it genuinely staggers me how everyone is afraid of criticizing this PR. I **really, really hate** how Skyrat downplays the importance of criticism. No matter how unconstructive or poorly worded, criticism is important, and this github and the skyrat discord have no kind of criticism at all. Good idea, terrible execution, i really hope it does not get merged.

What I'm saying is that this benefits antagonists only, (the little that it does benefit) and affects EVERYONE at the cost of antag satisfaction/"gimmicks" that are underbaked and do more harm than good, like this.

**Comment by optimumtact** [comment]
> I really, really hate how Skyrat downplays the importance of criticism. No matter how unconstructive or poorly worded, criticism is important, and this github and the skyrat discord have no kind of criticism at all.
Good idea, terrible execution, i really hope it does not get merged.

maybe if you were actually capable of including any real criticism of why this is a bad execution of the concept people might actually take you seriously, instead of you just railing about it like a moron

**Comment by ChungusGamer666** [comment]
> maybe if you were actually capable of including any real criticism of why this is a bad execution of the concept people might actually take you seriously, instead of you just railing about it like a moron

First off, this mostly applies to the implementation seem here, on TG i have very few complaints besides dumb code shit like not registering the mob login signal and entirely deleting the component on logout (I think it would be just fine to delete the hud component on logout and regenerate it on login).
But here dentbrain, why this PR sucks in a nutshell:
- Vision cone is way too damn wide, 90 works just fine.
- Why the hell does combat mode have ANY effect on FoV? This is counter intuitive and prevents people from you know, doing what this is supposed to cause, backstabbing (unless they have a gun).
- Adjacent mask is stupid and prevents backstabbing even without combat mode enabled, so it hinders the entire point of having an FOV system.
- Chat messages ignore the fact you're hidden by the shadow cone, and get displayed ike if your character had perfect 360 degree vision.
- Lying down does not disable the field of vision, nor rotate it, which looks a bit stupid (Absolutely nitpicking)

In summary, this implementation of FoV has all the downsides, but none of the upsides of versions i've seem implemented elsewhere.
